service/history/shard/context_impl.go

```diff
@@ -1610,6 +1614,15 @@ func (s *ContextImpl) transition(request contextRequest) error {
 				s.contextTaggedLogger.Warn("transition to acquired but no engine set")
 				return errInvalidTransition
 			}
+
+			engine, err := s.engineFuture.Get(s.lifecycleCtx)
+			if err != nil {
+				// this should not happen, already checked engineFuture is ready above
+				s.contextTaggedLogger.Warn("transition to acquired but unable to get engine", tag.Error(err))
+				return errInvalidTransition
+			}
+			s.notifyQueueProcessor(engine)
```

Comment:
sorry for the late review.

can `notifyQueueProcessor` block at all? does it take any other locks? this whole function is within `s.stateLock`, which is supposed to be a pretty low-level lock. can this call be moved to `acquireShard` instead so it runs outside of `stateLock`?

Reply:
oh good point! yeah, notifyQueueProcessor may try to get shard lock, it won't cause deadlock as they are two separate locks I guess. But I will see if I can move the logic to acquireShard.

Reply:
I think that is a potential deadlock: many methods here acquire state lock inside of shard lock. the intention is that state lock has no other locks inside it (except the future itself I guess)

Reply:
yeah, moved to acquireShard to avoid potential issues. Mind take a look?
https://github.com/temporalio/temporal/pull/3392